### PR TITLE
Mechanism for filtering linked elements visible in the active view of host document added

### LINKED ISSUE #1261: Enable pulling linked elements that are visible in the active view of the host document
#### Description:

Following @enarhi's request, this has been identified as a time-saving feature.

### PR DESCRIPTION
**Pull request by pawelbaran**
### Issues addressed by this PR

Closes #1261

### Test files

On [SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FRevit%5FToolkit%2F%231262%2DLinkedElementsFromActiveView&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4) - try hiding the elements to validate the behaviour.

### Additional comments

@vietle-bh @travispotterBH @michal-pekacki please have a look at `ActiveViewVisibilityContext` and its application in filtering the visible linked elements - can be helpful in other applications 👍

### CODE REVIEW

**Comment by pawelbaran** [comment]
@BHoMBot check core
@BHoMBot check compliance

**Comment by pawelbaran** [comment]
@BHoMBot check core
@BHoMBot check compliance

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Comment by pawelbaran** [comment]
@BHoMBot check versioning

**Comment by pawelbaran** [comment]
@BHoMBot check null-handling
@BHoMBot check serialisation

File: Revit_Core_Engine/Objects/ActiveViewVisibilityContext.cs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        {
+            return true;
+        }
+
+
+        /***************************************************/
+        /****              Private fields               ****/
+        /***************************************************/
+
+        private Document m_targetDocument;
```
[the remark this answers is not in the document]
Comment (pawelbaran) [reply Revit_Core_Engine/Objects/ActiveViewVisibilityContext.cs]: Good spot, fixed now 👍

**Comment by vietle-bh** [comment]
Just to confirm, this only works for 3D view, doesn't it?
I got this from plan views and sections:

image

**Comment by pawelbaran** [comment]
Which Revit version did you test against @vietle-bh? The API has changed in the context of applicable views, in older versions it was only 3d, but then I thought some more got enabled

**Comment by pawelbaran** [comment]
So... In Revit 2018 we have this:
image

...while in 2023...
image

So it should work for at least some types of 2d views I guess?

**Comment by pawelbaran** [comment]
Oke, I did some more testing and confirm that the feature seems not to work for 2d views in Revit 2023. But I tweaked the exception catching mechanism to automatically unlock other view types as soon as they start working. Can't think of how it could be improved further now 🙈

**Comment by vietle-bh** [comment]
Great 👍  Can't wait for the future!

Nevertheless, [the docs](https://www.revitapidocs.com/2023/d2437433-9183-cbb1-1c67-dedd86db5b5a.htm) seems to suggest using a slightly different interface for 2D views:
- _For 2D views, an instance of [IExportContext2D ](https://www.revitapidocs.com/2023/a4578846-6ecf-e354-668d-96d8ef5d1a32.htm)has to be used._

And there's a list of supported 2D view types [here](https://www.revitapidocs.com/2023/58d06458-fd6a-bdef-c457-2c52b50a70e8.htm).
- _the only 2D view types exported are [IncludeGeometricObjects ](https://www.revitapidocs.com/2023/2ce1075e-380e-01e7-6459-b7467c2a2414.htm)FloorPlan, CeilingPlan,Elevation, Section, Detail, EngineeringPlan, AreaPlan_

Not sure it's worth implementing though.

**Comment by pawelbaran** [comment]
> Great 👍 Can't wait for the future!
> 
> Nevertheless, [the docs](https://www.revitapidocs.com/2023/d2437433-9183-cbb1-1c67-dedd86db5b5a.htm) seems to suggest using a slightly different interface for 2D views:
> 
> * _For 2D views, an instance of [IExportContext2D ](https://www.revitapidocs.com/2023/a4578846-6ecf-e354-668d-96d8ef5d1a32.htm)has to be used._
> 
> And there's a list of supported 2D view types [here](https://www.revitapidocs.com/2023/58d06458-fd6a-bdef-c457-2c52b50a70e8.htm).
> 
> * _the only 2D view types exported are [IncludeGeometricObjects ](https://www.revitapidocs.com/2023/2ce1075e-380e-01e7-6459-b7467c2a2414.htm)FloorPlan, CeilingPlan,Elevation, Section, Detail, EngineeringPlan, AreaPlan_
> 
> Not sure it's worth implementing though.

You had perfect timing @vietle-bh! We were just discussing this part of the documentation with @enarhi when you added your comment. I just came back to the computer, got the `IExportContext2D` implemented. Unfortunately, I needed to create a separate class per each interface because if one class implements both, it only works on 2d views (3d throwing an error, dunno why 🐒). So now it all should work.

One last note on the topic of generalising the filtering mechanism: currently the context classes are tailored for querying the active view only, but I am pretty sure giving them a bit more love would enable filtering other views than the active one. This time, however, I took the yagni approach, happy to get the feature extended once we actually need it. FYI @travispotterBH @michal-pekacki

**Review by vietle-bh — COMMENTED** [review]
Now it also works for 2D views 👍 
Anyway, I noticed that the below method usually include the Revit link in the result even when that link is outside of the 2D or 3D view's crop region. Looks like a limitation of `CustomExporter`. Should we add a bounding box check for Revit links to fix this?

image

**Comment by pawelbaran** [comment]
> Now it also works for 2D views 👍 Anyway, I noticed that the below method usually includes the Revit link in the result even when that link is outside of the 2D or 3D view's crop region. Looks like a limitation of `CustomExporter`. Should we add a bounding box check for Revit links to fix this?
> 
> image

Do you mean the link itself @vietle-bh? Or the member elements of the link? If only the link, then I would leave it as it, usually filtering by active view will be combined with other queries I guess...

**Comment by vietle-bh** [comment]
> Do you mean the link itself @vietle-bh? Or the member elements of the link? If only the link, then I would leave it as it, usually filtering by active view will be combined with other queries I guess...

It's the Revit link instance itself, and yes, let's leave it as-is 👍

**Review by vietle-bh — DISMISSED** [review]
Testing and discussion went well 👍

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Comment by pawelbaran** [comment]
@BHoMBot check core

**Comment by pawelbaran** [comment]
@BHoMBot check core

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Comment by pawelbaran** [comment]
@BHoMBot check versioning

**Comment by pawelbaran** [comment]
@BHoMBot check versioning

**Comment by pawelbaran** [comment]
@BHoMBot check installer
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance

**Review by vietle-bh — APPROVED** [review]
There's only 1 commit after my last review and it didn't break anything 👍

**Pull request merged**